### DISCUSSION TITLE
How can I restrict users from accessing a client?

### DISCUSSION THREAD

**Discussion by mtrana** [keycloak/keycloak#24157] [Q&A]
Hello, 
i have the following problem:

I have two Applications : **App A** and **App B** 

_I have 3 roles:_ 
1. **role A** (should only have access to App A)
2. **role B** (should only have access to App B)
3. **role C** (should have access to App A and B)

My Idea: 

Create one Realm with two clients: 

-  **Client A** with client **role A** and,
-  **Client B** with client **role B** 

Create a Realm **role C**
And finally create three users and map each to one of the roles above.

WIth this I want to achieve that the user with the role A cannot have access to the client B but cllient C can have access to both clients.

**BUT** this doesnt work as expected because I can login with the **client role A** also to **client B**

Am I missing any configuration?  Am I on the right track? If not how can I solve this?

Thanks in advance!

**Comment by danielFesenmeyer** [comment]
As I understand, you are able to **authenticate** with `client B`, even when the client does not have `role B` in the scope - which is expected. But you should not have the `role B` in the (access) token, thus **authorization** should fail, if you try to run any actions secured with this role. Maybe having this security on **authorization** level is sufficient for you?

Otherwise, there is a third-party solution which applies client restriction on **authentication** (provided by @sventorben) https://github.com/sventorben/keycloak-restrict-client-auth
